src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java

```diff
@@ -1,26 +1,33 @@
 package pl.cleankod.exchange.provider;
 
+import io.github.resilience4j.cache.Cache;
 import pl.cleankod.exchange.core.domain.Money;
 import pl.cleankod.exchange.core.gateway.CurrencyConversionService;
 import pl.cleankod.exchange.provider.nbp.ExchangeRatesNbpClient;
 import pl.cleankod.exchange.provider.nbp.model.RateWrapper;
 
-import java.math.BigDecimal;
 import java.math.RoundingMode;
 import java.util.Currency;
 
 public class CurrencyConversionNbpService implements CurrencyConversionService {
     private final ExchangeRatesNbpClient exchangeRatesNbpClient;
+    private final Cache<String, RateWrapper> cache;
 
-    public CurrencyConversionNbpService(ExchangeRatesNbpClient exchangeRatesNbpClient) {
+    public CurrencyConversionNbpService(ExchangeRatesNbpClient exchangeRatesNbpClient, Cache<String, RateWrapper> cache) {
         this.exchangeRatesNbpClient = exchangeRatesNbpClient;
+        this.cache = cache;
     }
 
     @Override
     public Money convert(Money money, Currency targetCurrency) {
-        RateWrapper rateWrapper = exchangeRatesNbpClient.fetch("A", targetCurrency.getCurrencyCode());
-        BigDecimal midRate = rateWrapper.rates().get(0).mid();
-        BigDecimal calculatedRate = money.amount().divide(midRate, RoundingMode.HALF_UP);
+        var rateWrapper = Cache.decorateSupplier(
+                        cache, () -> exchangeRatesNbpClient.fetch("A", targetCurrency.getCurrencyCode())
+                )
+                .apply(targetCurrency.getCurrencyCode());
+        var midRate = rateWrapper.rates().get(0).mid();
+        var calculatedRate = money.amount()
+                .divide(midRate, RoundingMode.HALF_EVEN);
```

Comment:
maybe it would be better to perform all the $ related stuff in the `Money` class. Also, note that money calculation precision is > #.##